pkg/storage/unified/sql/batch.go

```diff
@@ -0,0 +1,328 @@
+		rv := int64(0)
+
+		summaries := make(map[string]*resource.BatchResponse_Summary, len(setting.Collection)*4)
+
+		// First clear everything in the transaction
+		if setting.RebuildCollection {
+			for _, key := range setting.Collection {
+				summary, err := batch.deleteCollection(key)
+				if err != nil {
+					return rollbackWithError(err)
+				}
+				summaries[key.BatchID()] = summary
+				rsp.Summary = append(rsp.Summary, summary)
+
+				// get the RV for this resource
+				tmp, err := b.resourceVersionAtomicInc(ctx, tx, key)
+				if err != nil {
+					return rollbackWithError(fmt.Errorf("increment resource version: %w", err))
+				}
+				if tmp > rv {
+					rv = tmp
+				}
+			}
+		}
+		if rv < 100 {
+			return rollbackWithError(fmt.Errorf("expected resource version to be set"))
+		}
+
+		obj := &unstructured.Unstructured{}
+
+		// Write each event into the history
+		for iter.Next() {
+			if iter.RollbackRequested() {
+				return rollbackWithError(nil)
+			}
+			req := iter.Request()
+			if req == nil {
+				return rollbackWithError(fmt.Errorf("missing request"))
+			}
+			rsp.Processed++
+
+			if req.Action == resource.BatchRequest_UNKNOWN {
+				rsp.Rejected = append(rsp.Rejected, &resource.BatchResponse_Rejected{
+					Key:    req.Key,
+					Action: req.Action,
+					Error:  "unknown action",
+				})
+				continue
+			}
+
+			err := obj.UnmarshalJSON(req.Value)
+			if err != nil {
+				rsp.Rejected = append(rsp.Rejected, &resource.BatchResponse_Rejected{
+					Key:    req.Key,
+					Action: req.Action,
+					Error:  "unable to unmarshal json",
+				})
+				continue
+			}
+
+			rv++ // Increment the resource version
+
+			// Write the event to history
+			if _, err := dbutil.Exec(ctx, tx, sqlResourceHistoryInsert, sqlResourceRequest{
+				SQLTemplate: sqltemplate.New(b.dialect),
+				WriteEvent: resource.WriteEvent{
+					Key:        req.Key,
+					Type:       resource.WatchEvent_Type(req.Action),
+					Value:      req.Value,
+					PreviousRV: -1, // Used for WATCH, but we want to skip watch events
+				},
+				Folder:          req.Folder,
+				GUID:            uuid.NewString(),
+				ResourceVersion: rv,
+			}); err != nil {
+				return rollbackWithError(fmt.Errorf("insert into resource history: %w", err))
+			}
+
+			// fmt.Printf("WROTE %4d/%d/%s\n", rsp.Processed, rv, req.Key.Name)
+		}
+
+		// Now update the resource table from history
+		for _, key := range setting.Collection {
+			k := fmt.Sprintf("%s/%s/%s", key.Namespace, key.Group, key.Resource)
+			summary := summaries[k]
+			if summary == nil {
+				return rollbackWithError(fmt.Errorf("missing summary key for: %s", k))
+			}
+
+			err := batch.syncCollection(key, summary)
+			if err != nil {
+				return err
+			}
+
+			// Make sure the collection RV is above our last written event
+			for {
+				crv, _ := b.resourceVersionAtomicInc(ctx, tx, key)
+				if crv > rv {
+					break
+				}
+				time.Sleep(10 * time.Millisecond)
+			}
+		}
```

Comment:
Given we are writing history, shall we generate the ResourceVersion based on the creation timestamp instead  ?  Using resourceVersionAtomicInc is probably a big bottleneck.